libraries/standard/mqtt/src/mqtt.c

```diff
@@ -99,6 +101,509 @@ static int32_t sendPacket( MQTTContext_t * pContext,
+ *
+ * @param[in] pContext Initialized MQTT Context.
+ *
+ * @return #MQTTKeepAliveTimeout if a PINGRESP is not received in time,
+ * #MQTTSendFailed if the PINGREQ cannot be sent, or #MQTTSuccess.
+ */
+static MQTTStatus_t handleKeepAlive( MQTTContext_t * const pContext )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    uint32_t now = 0U, keepAliveMs = 0U;
+
+    assert( pContext != NULL );
+    now = pContext->callbacks.getTime();
+    keepAliveMs = 1000U * ( uint32_t ) pContext->keepAliveIntervalSec;
+
+    /* If keep alive interval is 0, it is disabled. */
+    if( ( keepAliveMs != 0U ) && 
+        ( calculateElapsedTime( now, pContext->lastPacketTime ) > keepAliveMs ) )
+    {
+        if( pContext->waitingForPingResp )
+        {
+            /* Has time expired? */
+            if( calculateElapsedTime( now, pContext->pingReqSendTimeMs ) > 
+                pContext->pingRespTimeoutMs )
+            {
+                status = MQTTKeepAliveTimeout;
+            }
+        }
+        else
+        {
+            status = MQTT_Ping( pContext );
+            /* Sending the packet may have taken some time. Recalculate. */
+            pContext->pingReqSendTimeMs = pContext->callbacks.getTime();
+            pContext->waitingForPingResp = true;
+        }
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Handle received MQTT PUBLISH packet.
+ *
+ * @param[in] pContext MQTT Connection context.
+ * @param[in] pIncomingPacket Incoming packet.
+ *
+ * @return MQTTSuccess, MQTTIllegalState or deserialization error.
+ */
+static MQTTStatus_t handleIncomingPublish( MQTTContext_t * const pContext,
+                                           MQTTPacketInfo_t * pIncomingPacket )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTPublishState_t publishRecordState = MQTTStateNull;
+    uint16_t packetIdentifier;
+    MQTTPublishInfo_t publishInfo;
+
+    assert( pContext != NULL );
+    assert( pIncomingPacket != NULL );
+
+    status = MQTT_DeserializePublish( pIncomingPacket, &packetIdentifier, &publishInfo );
+    IotLogInfoWithArgs( "De-serialized incoming PUBLISH packet: DeserializerResult=%d", status );
+
+    if( status == MQTTSuccess )
+    {
+        publishRecordState = MQTT_UpdateStatePublish( pContext,
+                                                      packetIdentifier,
+                                                      MQTT_RECEIVE,
+                                                      publishInfo.qos );
+        IotLogInfoWithArgs( "State record updated. New state=%d.",
+                            publishRecordState );
+
+        /* Send PUBACK or PUBREC if necessary. */
+        status = sendPublishAcks( pContext,
+                                  packetIdentifier,
+                                  &publishRecordState );
+    }
+
+    if( status == MQTTSuccess )
+    {
+        /* TODO: Should the publish info be passed instead? */
```

Comment:
What is the benefit with that?

Reply:
All of the information related to the publish is stored in the MQTTPublishInfo_t, including QoS, topic name, and message payload. The only thing the packet info includes is the packet type (PUBLISH + flags), and a pointer to the serialized data.

Reply:
I also think then should definitely pass the publish info rather than the packet info so as the application can make use of the already de-serialized data.

Reply:
What do we do for the ack callbacks ? We are only using one callback for reporting all incoming events.

Reply:
The issue is that MQTT_DeserializePublish() does not change `pIncomingPacket` in any way and stores the information in the publish info. If the application wants the deserialized publish information then it will need to call MQTT_DeserializePublish() again, which is wasteful since the data has already been deserialized and is available.